### PR TITLE
avm2: Convert filters to ActionScript

### PR DESCRIPTION
**Pull request by Aaron1011**
We still don't implement the actual filter logic for `DisplayObject`,
but the classes themselves are fully implemented.

### CODE REVIEW

File: core/src/avm2/globals/stubs.as [hunk 1]
```diff
@@ -5,3 +5,6 @@ include "Object.as"
 include "Boolean.as"
 include "Number.as"
 include "String.as"
+include "Array.as"
```
Comment (relrelb) [inline core/src/avm2/globals/stubs.as:8]: Let's sort this list alphabetically (might also be worth to mention it in a comment).

File: core/src/avm2/globals/globals.as [hunk 2]
```diff
@@ -1,3 +1,7 @@
 include "flash/geom/Point.as"
 include "flash/geom/Rectangle.as"
 include "flash/text/AntiAliasType.as"
+include "flash/filters/BitmapFilter.as"
```
Comment (relrelb) [inline core/src/avm2/globals/globals.as:4]: Let's sort this list alphabetically (might also be worth to mention it in a comment).

File: core/src/avm2/globals/flash/filters/BlurFilter.as [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+		public var quality: int;
+
+		public function BlurFilter(blurX: Number = 4.0, blurY: Number = 4.0, quality: int = 1) {
+			this.blurX = blurX;
+			this.blurY = blurY;
+			this.quality = qualty;
+		}
+
+		override public function clone(): BitmapFilter {
+			return new BitmapFilter(this.blurX, this.blurY, this.quality);
```
Comment (relrelb) [inline core/src/avm2/globals/flash/filters/BlurFilter.as:14]: Have you meant `new BlurFilter`?

File: core/src/avm2/globals/flash/filters/ColorMatrixFilter.as [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿package flash.filters {
+	public final class ColorMatrixFilter extends BitmapFilter {
+		public var matrix: Array;
+
+		public function ColorMatrixFilter(matrix: Array = null) {
+			this.matrix = matrix;
+		}
+		override public function clone(): BitmapFilter {
+			return new ColorMatrixFilter(this.matrix.concat());
```
Comment (relrelb) [inline core/src/avm2/globals/flash/filters/ColorMatrixFilter.as:9]: Why `.concat()`?

File: core/src/avm2/globals/flash/filters/BlurFilter.as (hunk 3 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/filters/BlurFilter.as]: Fixed - it seems that `asc.jar` doesn't attempt to resolve the constructor call.

**Comment by relrelb** [comment]
I think we should wait for #7244 to be merged first, as I see that the `flash.filters.*` classes have native getters/setters, rather than regular ActionScript properties.

File: core/src/avm2/globals/flash/filters/ColorMatrixFilter.as (hunk 4 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/filters/ColorMatrixFilter.as:9]: Calling `concat()` makes a shallow copy (which is fine here, since `this.matrix` should contain numbers). Otherwise, we'd end up with two `ColorMatrixFilters` with the same underlying array.

**Comment by Aaron1011** [comment]
At the moment, there's no need for native getters/setters (and when we do implement the actual filter logic, we could always choose to grab the matrix from the filter on the Rust side). However, the getters/setters *are* necessary to make a copy of the `Array` (the docs state that changing the returned `Array` directly has no effect). I've updated the PR accordingly.

**Comment by danielhjacobs** [comment]
Not directly related or necessarily to be part of this PR but I think I noticed that The New York Times animations that are not stuck on timer are stuck on drop shadow filter which is pretty similar to these filters.  I'm not sure what they'll be stuck on once that's implemented but just worth keeping in mind.

File: core/src/avm2/globals/flash/filters/ColorMatrixFilter.as (hunk 4 above)
Comment (relrelb) [reply core/src/avm2/globals/flash/filters/ColorMatrixFilter.as:21]: This is still not 100% accurate, since `.matrix` should be an array of 20 items for 4 x 5 color transform matrix (See [documentation](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/filters/ColorMatrixFilter.html#matrix)).

Consider the following AS3 code:
```as
import flash.filters.*;
var a = [1, 2, 3];
var f = new ColorMatrixFilter(a);
trace(f.matrix); // traces "1,2,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0" - an array of 20 items
```

Also, note that arrays containing more than 20 items are clamped:
```as
import flash.filters.*;
var a = [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,3];
var f = new ColorMatrixFilter(a);
trace(f.matrix); // traces "1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2" - an array of 20 items
```

And that's not all - The items should be coerced to `f64` upon a `ColorMatrixFilter` creation:
```as
import flash.filters.*;
var a = ["1.5"];
var f = new ColorMatrixFilter(a);
trace(typeof f.matrix[0]); // traces "number" (which is 1.5)
```

In short, it's very likely that Flash stored the matrix data in `[f64; 20]` (or even better, `[[f64; 5]; 4]`), so I highly recommend doing so in Ruffle too, once we have native bindings (#7244).

**Review by Herschel — APPROVED** [review]
Thank you!
